README
Unsyntax is an expander for R7RS programs.  This package contains an
interpreter, a compiler, and a REPL.  It currently uses Chibi-Scheme as a
backend.

Unsyntax's home page is https://www.unsyntax.org

* Installation

** Build from git

The README-hacking is about building, modifying and checking Unsyntax.  See
its "Working from the Repository" section to build Unsyntax from the git
repo.  Roughly, run:

    $ ./bootstrap

then proceed with the usual ‘configure && make’ steps.

** Build from tarball

See the file INSTALL for generic compilation and installation instructions.

Unsyntax requires (the development version of) Chibi-Scheme to build and
run.

* Bug reports

Please send bug reports to <marc@nieper-wisskirchen.de>.  Be sure to include
the version number from ‘unsyntax-scheme --version’, and a complete,
self-contained test case in each bug report.

* Copyright statements

For any copyright year range specified as YYYY-ZZZZ in this package, note
that the range specifies every single year in that closed interval.

-----

Copyright (C) 2020 Marc Nieper-Wißkirchen

This file is part of Unsyntax.

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation files
(the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice (including the
next paragraph) shall be included in all copies or substantial
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

Local Variables:
mode: outline
fill-column: 76
End:


NEWS
Unsyntax NEWS

* Noteworthy changes in release ?.? (????-??-??) [?]

** New features

*** SRFI 27

  SRFI 27 is now supported.

*** Records

  The whole R6RS record library has been implemented on top of R7RS
  records.

*** Meta Definitions

  The new library (unsyntax) exports the ‘meta’ keyword that turns
  forms into meta definitions.  A meta definition is evaluated at
  expand-time and available to the right-hand sides of syntax and
  property definitions.

*** Modules and local imports

  Libraries can now be lexically locally imported.  Moreover, a new
  module form allows to define modules in a lexically local scope.

** Libraries

*** SRFI 111

  SRFI 111 can now also be imported under its R7RS (large) name
  (scheme box).

** Bug fixes

*** Reader

  The reader now reads symbols enclosed in ‘|’.

* Noteworthy changes in release 0.0.3 (2020-10-18) [alpha]

** User interface

*** When a keyword is used, to which no proper macro transformer is
    associated, the source location of the define-syntax is now
    reported.

** New features

*** SRFI 111

  SRFI 111 has been made available.

*** SRFI 211

  All procedures and syntax of SRFI 211 is now supported.  This
  includes support for explicit and implicit renaming macros, the
  syntactic closure macro facility, and old-style Lisp macros.

*** SRFI 213

  Identifier properties from SRFI 213 are now supported.

*** R6RS compatibility

  Unsyntax now supports R6RS library declarations.

  Library names of the form (srfi nnn) are totally equivalent to those
  of the form (srfi :nnn).

  Recognized extensions for library files are ‘.unsyntax.sls’, ‘.sls’,
  and ‘.sld’.

*** R7RS conformance

  Unsyntax now supports multiple import declarations at the beginning
  of a program.

** Various bug fixes

* Noteworthy changes in release 0.0.2 (2020-09-24) [alpha]

** Bug fixes

*** Interactive environment

  The REPL and the interactive environment are now complete.  In
  particular, they now allow top-level definitions.

** New features

*** SRFI 188

  SRFI 188 (Splicing binding constructs for syntactic keywords) is now
  supported.

*** SRFI 206

  SRFI 206 (Auxiliary Syntax Keywords) is now supported.

*** SRFI 212

  SRFI 212 (Aliases) is now supported.

** Documentation

  Man pages and a manual have been added.

* Noteworthy changes in release 0.0.1 (2020-09-17) [alpha]

  Initial release.

-----

Copyright (C) 2020 Marc Nieper-Wißkirchen

This file is part of Unsyntax.

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation files
(the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice (including the
next paragraph) shall be included in all copies or substantial
portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


src/stdlibs.scm
;; Copyright © Marc Nieper-Wißkirchen (2020).

;; This file is part of Unsyntax.

;; Permission is hereby granted, free of charge, to any person
;; obtaining a copy of this software and associated documentation files
;; (the "Software"), to deal in the Software without restriction,
;; including without limitation the rights to use, copy, modify, merge,
;; publish, distribute, sublicense, and/or sell copies of the Software,
;; and to permit persons to whom the Software is furnished to do so,
;; subject to the following conditions:

;; The above copyright notice and this permission notice (including the
;; next paragraph) shall be included in all copies or substantial
;; portions of the Software.

;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
;; NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
;; BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
;; ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
;; CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
;; SOFTWARE.

(rnrs records syntactic)
(rnrs records procedural)
(rnrs records inspection)
(scheme base)
(scheme case-lambda)
(scheme char)
(scheme complex)
(scheme cxr)
(scheme eval)
(scheme file)
(scheme inexact)
(scheme lazy)
(scheme load)
(scheme process-context)
(scheme read)
(scheme repl)
(scheme r5rs)
(scheme time)
(scheme write)
(srfi 1)
(srfi 2)
(srfi 8)
(srfi 27)
(srfi 28)
(srfi 37)
(srfi 59)
(srfi 64)
(srfi 111)
(srfi 125)
(srfi 128)
(srfi 139)
(srfi 158)
(srfi 188)
(srfi 190)
(srfi 206)
(srfi 211 define-macro)
(srfi 211 explicit-renaming)
(srfi 211 identifier-syntax)
(srfi 211 implicit-renaming)
(srfi 211 low-level)
(srfi 211 syntactic-closures)
(srfi 211 syntax-case)
(srfi 211 syntax-parameter)
(srfi 211 variable-transformer)
(srfi 211 with-ellipsis)
(srfi 212)
(srfi 213)
(unsyntax)
(alias (srfi 1) (scheme list))
(alias (srfi 111) (scheme box))
(alias (srfi 125) (scheme hash-table))
(alias (srfi 128) (scheme comparator))
(auxiliary-syntax (srfi 206 all))
